File: hope_ws/src/calibration/src/split_calibration_csv.cpp
#include "split_calibration_csv.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <set>
#include <sstream>
#include <stdexcept>
#include <string>

namespace calibration {

namespace {

double median(std::vector<double> v) {
  if (v.empty()) {
    return 0.0;
  }
  std::sort(v.begin(), v.end());
  size_t n = v.size();
  if (n % 2 == 0) {
    return 0.5 * (v[n / 2 - 1] + v[n / 2]);
  }
  return v[n / 2];
}

std::set<int> parseSelectedIds(const std::string & text) {
  std::set<int> out;
  if (text.empty()) {
    return out;
  }
  std::stringstream ss(text);
  std::string part;
  while (std::getline(ss, part, ',')) {
    auto first = part.find_first_not_of(" \t");
    auto last = part.find_last_not_of(" \t");
    if (first == std::string::npos) {
      continue;
    }
    out.insert(std::stoi(part.substr(first, last - first + 1)));
  }
  return out;
}

std::vector<std::pair<double, Eigen::Vector3d>> loadCsv(const std::string & path) {
  std::ifstream fh(path);
  if (!fh) {
    throw std::runtime_error("Cannot open CSV: " + path);
  }
  std::vector<std::pair<double, Eigen::Vector3d>> rows;
  std::string line;
  while (std::getline(fh, line)) {
    if (line.empty()) {
      continue;
    }
    std::stringstream ss(line);
    std::string tok;
    std::vector<std::string> cols;
    while (std::getline(ss, tok, ',')) {
      cols.push_back(tok);
    }
    if (cols.size() < 4) {
      continue;
    }
    std::string first = cols[0];
    std::transform(first.begin(), first.end(), first.begin(),
      [](unsigned char c) { return std::tolower(c); });
    if (first == "t" || first == "time") {
      continue;
    }
    double t = std::stod(cols[0]);
    Eigen::Vector3d p(std::stod(cols[1]), std::stod(cols[2]), std::stod(cols[3]));
    rows.emplace_back(t, p);
  }
  if (rows.empty()) {
    throw std::runtime_error("No numeric rows found in CSV: " + path);
  }
  return rows;
}

void writeManifest(
  const std::string & path,
  const std::vector<ChunkSummary> & manifests)
{
  std::ofstream mf(path);
  if (!mf) {
    throw std::runtime_error("Cannot write manifest.csv: " + path);
  }
  mf << "chunk_id,rows,duration_s,scale_to_m,z_offset_m,"
        "x_min_m,x_max_m,y_min_m,y_max_m,z_min_m,z_max_m\n";
  mf << std::fixed << std::setprecision(6);
  for (const auto & s : manifests) {
    mf << s.chunk_id << "," << s.rows << "," << s.duration_s << ","
       << s.scale_to_m << "," << s.z_offset_m << ","
       << s.x_min_m << "," << s.x_max_m << ","
       << s.y_min_m << "," << s.y_max_m << ","
       << s.z_min_m << "," << s.z_max_m << "\n";
  }
}

ChunkSummary summarize(
  int chunk_id,
  const std::vector<double> & times,
  const std::vector<Eigen::Vector3d> & pos,
  double scale_to_m,
  double z_offset_m)
{
  ChunkSummary s;
  s.chunk_id = chunk_id;
  s.rows = static_cast<int>(times.size());
  s.duration_s = (times.size() > 1) ? times.back() : 0.0;
  s.scale_to_m = scale_to_m;
  s.z_offset_m = z_offset_m;
  if (!pos.empty()) {
    s.x_min_m = pos[0].x(); s.x_max_m = pos[0].x();
    s.y_min_m = pos[0].y(); s.y_max_m = pos[0].y();
    s.z_min_m = pos[0].z(); s.z_max_m = pos[0].z();
    for (const auto & p : pos) {
      s.x_min_m = std::min(s.x_min_m, p.x()); s.x_max_m = std::max(s.x_max_m, p.x());
      s.y_min_m = std::min(s.y_min_m, p.y()); s.y_max_m = std::max(s.y_max_m, p.y());
      s.z_min_m = std::min(s.z_min_m, p.z()); s.z_max_m = std::max(s.z_max_m, p.z());
    }
  }
  return s;
}

}  // namespace

double inferScale(const std::vector<Eigen::Vector3d> & pos) {
  if (pos.empty()) {
    return 1.0;
  }
  std::vector<double> coords;
  coords.reserve(pos.size() * 3);
  for (const auto & p : pos) {
    coords.push_back(std::abs(p.x()));
    coords.push_back(std::abs(p.y()));
    coords.push_back(std::abs(p.z()));
  }
  double median_abs = median(coords);
  return (median_abs > 10.0) ? 0.001 : 1.0;
}

std::vector<std::vector<Eigen::Vector3d>> splitRows(
  const std::vector<double> & times,
  const std::vector<Eigen::Vector3d> & pos,
  double gap_seconds)
{
  std::vector<std::vector<Eigen::Vector3d>> chunks;
  if (times.size() <= 1) {
    chunks.push_back(pos);
    return chunks;
  }
  size_t start = 0;
  for (size_t i = 0; i + 1 < times.size(); ++i) {
    if (times[i + 1] - times[i] > gap_seconds) {
      std::vector<Eigen::Vector3d> chunk(pos.begin() + start, pos.begin() + i + 1);
      chunks.push_back(chunk);
      start = i + 1;
    }
  }
  std::vector<Eigen::Vector3d> last(pos.begin() + start, pos.end());
  chunks.push_back(last);
  return chunks;
}

NormalizedChunk normalizeChunk(
  const std::vector<double> & times,
  const std::vector<Eigen::Vector3d> & pos,
  double scale_to_m)
{
  NormalizedChunk out;
  if (times.empty()) {
    return out;
  }
  out.times.reserve(times.size());
  out.pos.reserve(pos.size());
  double t0 = times.front();
  for (size_t i = 0; i < times.size(); ++i) {
    out.times.push_back(times[i] - t0);
    out.pos.push_back(pos[i] * scale_to_m);
  }
  std::vector<double> zs;
  zs.reserve(out.pos.size());
  for (const auto & p : out.pos) {
    zs.push_back(p.z());
  }
  std::sort(zs.begin(), zs.end());
  size_t low_count = std::max<size_t>(3, zs.size() / 10);
  double z_offset = 0.0;
  for (size_t i = 0; i < low_count; ++i) {
    z_offset += zs[i];
  }
  z_offset /= static_cast<double>(low_count);
  out.z_offset_m = z_offset;
  for (auto & p : out.pos) {
    p.z() -= z_offset;
  }
  return out;
}

void writeChunkCsv(
  const std::string & output_dir,
  const std::string & prefix,
  int chunk_id,
  const std::vector<double> & times,
  const std::vector<Eigen::Vector3d> & pos)
{
  char fname[512];
  std::snprintf(fname, sizeof(fname), "%s/%s_chunk%02d.csv",
    output_dir.c_str(), prefix.c_str(), chunk_id);
  std::ofstream fh(fname);
  if (!fh) {
    throw std::runtime_error("Cannot write chunk CSV: " + std::string(fname));
  }
  fh << "t,x,y,z\n";
  fh << std::fixed << std::setprecision(9);
  for (size_t i = 0; i < times.size(); ++i) {
    fh << times[i] << "," << pos[i].x() << "," << pos[i].y() << "," << pos[i].z() << "\n";
  }
}

int splitCalibrationCsv(const SplitOptions & opts) {
  auto rows = loadCsv(opts.input);
  std::vector<double> times;
  std::vector<Eigen::Vector3d> pos;
  times.reserve(rows.size());
  pos.reserve(rows.size());
  for (const auto & [t, p] : rows) {
    times.push_back(t);
    pos.push_back(p);
  }
  double scale = inferScale(pos);
  std::set<int> selected = parseSelectedIds(opts.select);

  auto chunks = splitRows(times, pos, opts.gap_seconds);
  std::vector<ChunkSummary> manifests;
  int written = 0;
  for (size_t i = 0; i < chunks.size(); ++i) {
    int chunk_id = static_cast<int>(i + 1);
    auto norm = normalizeChunk(times, pos, scale);
    auto sum = summarize(chunk_id, norm.times, norm.pos, scale, norm.z_offset_m);
    manifests.push_back(sum);
    if (!selected.empty() && selected.count(chunk_id) == 0) {
      continue;
    }
    writeChunkCsv(opts.output_dir, opts.prefix, chunk_id, norm.times, norm.pos);
    written++;
  }

  writeManifest(opts.output_dir + "/manifest.csv", manifests);
  std::cout << "Input rows: " << rows.size() << "\n";
  std::cout << "Inferred scale_to_m: " << scale << "\n";
  std::cout << "Detected chunks: " << manifests.size() << "\n";
  std::cout << "Wrote chunk CSVs: " << written << "\n";
  std::cout << "Manifest: " << opts.output_dir << "/manifest.csv\n";
  return 0;
}

}  // namespace calibration

File: hope_ws/src/hope_planner/hope_planner/split_calibration_csv.py
"""Split a mixed calibration CSV into per-trajectory HOPE calibration CSVs.

This helper is for salvage situations where multiple tosses were recorded into
one file, the position unit is not yet meters, and the table-height zero point
is offset. It:

1. loads one `t,x,y,z` CSV,
2. infers whether the position unit is millimeters,
3. splits the recording on large time gaps, and
4. writes one normalized CSV per chunk with time relative to the chunk start
   and z shifted so the local low-z band is near the table plane.
"""

from __future__ import annotations

import argparse
import csv
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

import numpy as np


@dataclass(frozen=True)
class ChunkSummary:
    chunk_id: int
    rows: int
    duration_s: float
    scale_to_m: float
    z_offset_m: float
    x_min_m: float
    x_max_m: float
    y_min_m: float
    y_max_m: float
    z_min_m: float
    z_max_m: float


def _load_csv(path: Path) -> list[tuple[float, float, float, float]]:
    rows: list[tuple[float, float, float, float]] = []
    with path.open(newline="") as fh:
        reader = csv.reader(fh)
        for row in reader:
            if not row or row[0].strip().lower() in ("t", "time"):
                continue
            rows.append((float(row[0]), float(row[1]), float(row[2]), float(row[3])))
    if not rows:
        raise RuntimeError(f"No numeric rows found in {path}")
    return rows


def _infer_scale(rows: Iterable[tuple[float, float, float, float]]) -> float:
    coords = [abs(v) for _, x, y, z in rows for v in (x, y, z)]
    median_abs = float(np.median(coords))
    return 0.001 if median_abs > 10.0 else 1.0


def _split_rows(
    rows: list[tuple[float, float, float, float]],
    gap_seconds: float,
) -> list[list[tuple[float, float, float, float]]]:
    if len(rows) == 1:
        return [rows]

    chunks: list[list[tuple[float, float, float, float]]] = []
    start = 0
    for i in range(len(rows) - 1):
        if rows[i + 1][0] - rows[i][0] > gap_seconds:
            chunks.append(rows[start : i + 1])
            start = i + 1
    chunks.append(rows[start:])
    return chunks


def _normalize_chunk(
    rows: list[tuple[float, float, float, float]],
    scale_to_m: float,
) -> tuple[np.ndarray, np.ndarray, float]:
    raw = np.asarray(rows, dtype=float)
    t = raw[:, 0]
    pos = raw[:, 1:] * scale_to_m

    z_sorted = np.sort(pos[:, 2])
    low_count = max(3, len(z_sorted) // 10)
    z_offset_m = float(np.mean(z_sorted[:low_count]))
    pos[:, 2] -= z_offset_m

    return t - t[0], pos, z_offset_m


def _summarize_chunk(
    chunk_id: int,
    times: np.ndarray,
    pos: np.ndarray,
    scale_to_m: float,
    z_offset_m: float,
) -> ChunkSummary:
    return ChunkSummary(
        chunk_id=chunk_id,
        rows=len(times),
        duration_s=float(times[-1]) if len(times) > 1 else 0.0,
        scale_to_m=scale_to_m,
        z_offset_m=z_offset_m,
        x_min_m=float(np.min(pos[:, 0])),
        x_max_m=float(np.max(pos[:, 0])),
        y_min_m=float(np.min(pos[:, 1])),
        y_max_m=float(np.max(pos[:, 1])),
        z_min_m=float(np.min(pos[:, 2])),
        z_max_m=float(np.max(pos[:, 2])),
    )


def _write_chunk_csv(path: Path, times: np.ndarray, pos: np.ndarray) -> None:
    with path.open("w", newline="") as fh:
        writer = csv.writer(fh)
        writer.writerow(["t", "x", "y", "z"])
        for t, (x, y, z) in zip(times, pos):
            writer.writerow([f"{t:.9f}", f"{x:.9f}", f"{y:.9f}", f"{z:.9f}"])


def _parse_selected_ids(text: str | None) -> set[int] | None:
    if not text:
        return None
    out: set[int] = set()
    for part in text.split(","):
        stripped = part.strip()
        if not stripped:
            continue
        out.add(int(stripped))
    return out


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(
        description="Split one mixed t,x,y,z calibration CSV into per-trajectory CSVs."
    )
    parser.add_argument("--input", required=True, help="Input CSV path.")
    parser.add_argument("--output-dir", required=True, help="Directory for split CSVs.")
    parser.add_argument(
        "--gap-seconds",
        type=float,
        default=1.0,
        help="Start a new chunk when the timestamp gap exceeds this value.",
    )
    parser.add_argument(
        "--select",
        default="",
        help="Optional comma-separated chunk ids to export. Default: export every chunk.",
    )
    parser.add_argument(
        "--prefix",
        default="traj",
        help="Output file prefix. Example: traj -> traj_chunk02.csv",
    )
    args = parser.parse_args(argv)

    input_path = Path(args.input)
    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    rows = _load_csv(input_path)
    scale_to_m = _infer_scale(rows)
    selected_ids = _parse_selected_ids(args.select)

    manifest_rows: list[ChunkSummary] = []
    for chunk_id, chunk in enumerate(_split_rows(rows, args.gap_seconds), start=1):
        times, pos, z_offset_m = _normalize_chunk(chunk, scale_to_m)
        summary = _summarize_chunk(chunk_id, times, pos, scale_to_m, z_offset_m)
        manifest_rows.append(summary)

        if selected_ids is not None and chunk_id not in selected_ids:
            continue

        filename = f"{args.prefix}_chunk{chunk_id:02d}.csv"
        _write_chunk_csv(output_dir / filename, times, pos)

    manifest_path = output_dir / "manifest.csv"
    with manifest_path.open("w", newline="") as fh:
        writer = csv.writer(fh)
        writer.writerow([
            "chunk_id",
            "rows",
            "duration_s",
            "scale_to_m",
            "z_offset_m",
            "x_min_m",
            "x_max_m",
            "y_min_m",
            "y_max_m",
            "z_min_m",
            "z_max_m",
        ])
        for item in manifest_rows:
            writer.writerow([
                item.chunk_id,
                item.rows,
                f"{item.duration_s:.6f}",
                f"{item.scale_to_m:.6f}",
                f"{item.z_offset_m:.6f}",
                f"{item.x_min_m:.6f}",
                f"{item.x_max_m:.6f}",
                f"{item.y_min_m:.6f}",
                f"{item.y_max_m:.6f}",
                f"{item.z_min_m:.6f}",
                f"{item.z_max_m:.6f}",
            ])

    written = len(manifest_rows) if selected_ids is None else len(selected_ids)
    print(f"Input rows: {len(rows)}")
    print(f"Inferred scale_to_m: {scale_to_m}")
    print(f"Detected chunks: {len(manifest_rows)}")
    print(f"Wrote chunk CSVs: {written}")
    print(f"Manifest: {manifest_path}")


if __name__ == "__main__":
    main()
